src/node/miner.cpp

```diff
@@ -33,12 +33,10 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
     int64_t nOldTime = pblock->nTime;
     int64_t nNewTime{std::max<int64_t>(pindexPrev->GetMedianTimePast() + 1, TicksSinceEpoch<std::chrono::seconds>(NodeClock::now()))};
 
-    if (consensusParams.enforce_BIP94) {
-        // Height of block to be mined.
-        const int height{pindexPrev->nHeight + 1};
-        if (height % consensusParams.DifficultyAdjustmentInterval() == 0) {
-            nNewTime = std::max<int64_t>(nNewTime, pindexPrev->GetBlockTime() - MAX_TIMEWARP);
-        }
+    // Height of block to be mined.
+    const int height{pindexPrev->nHeight + 1};
+    if (height % consensusParams.DifficultyAdjustmentInterval() == 0) {
+        nNewTime = std::max<int64_t>(nNewTime, pindexPrev->GetBlockTime() - MAX_TIMEWARP);
```

Comment:
So why not drop `- MAX_TIMEWARP` and just make sure that we never propose a block with an earlier timestamp?

Or go beyond that and never allow time to go backwards on _any_ block?

Reply:
I finally wrapped my head around the purpose of `- MAX_TIMEWARP`. It's in fact very important that we use it here and also that we don't reduce it.

The [original Great Consensus Cleanup soft fork proposal](https://github.com/TheBlueMatt/bips/blob/cleanup-softfork/bip-XXXX.mediawiki) by @TheBlueMatt says the following:

> * Sadly, some deployed mining hardware relies on the ability to roll nTime forward by up to 600 seconds[3]. Thus, only requiring that the nTime field move forward during difficulty adjustment would allow a malicious miner to prevent some competitors from mining the next block by setting their timestamp to two hours in the future. Thus, we allow nTime to go backwards by 600 seconds, ensuring that even a block with a timestamp two hours in the future allows for 600 seconds of nTime rolling on the next block.

The only way to protect this specific mining hardware is by us setting `nTime` no higher than 2 hours minus 10 minutes in the future. We should add a comment to that effect here.

I also think we should increase this number (can't do that here of course, it's testnet4 consensus): https://delvingbitcoin.org/t/timewarp-attack-600-second-grace-period/1326

Reply:
Yes. So i say let's merge this and if we end up wanting to use a larger `MAX_TIMEWARP` as part of a mainnet soft fork, we'll be covered with the tighter value here.

Should i just add the comment as you suggested and we're good here?

Reply:
Let's wait to see if others agree with me on the Delving post that `600` is indeed too small, or at least that smaller would be bad. If so then indeed this should be fine to merge.

If we later increase `MAX_TIMEWARP` _and activate the soft fork_ then miners with timestamp rolling firmware should upgrade their Bitcoin Core version. Which they should do anyway if a timewarp fix is deployed, since the usual protection provided by standardness doesn't apply for that soft fork.